services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceRolesCondition.java

```diff
@@ -0,0 +1,134 @@
+
+    private static final Logger logger = Logger.getLogger(ClientUpdateSourceRolesCondition.class);
+
+    private final KeycloakSession session;
+    private final ComponentModel componentModel;
+
+    public ClientUpdateSourceRolesCondition(KeycloakSession session, ComponentModel componentModel) {
+        this.session = session;
+        this.componentModel = componentModel;
+    }
+
+    @Override
+    public String getName() {
+        return componentModel.getName();
+    }
+
+    @Override
+    public String getProviderId() {
+        return componentModel.getProviderId();
+    }
+
+    @Override
+    public ClientPolicyVote applyPolicy(ClientPolicyContext context) throws ClientPolicyException {
+        switch (context.getEvent()) {
+        case REGISTER:
+            if (context instanceof AdminClientRegisterContext) {
+                return getVoteForRolesMatched(((ClientUpdateContext)context).getAuthenticatedUser());
+            } else if (context instanceof DynamicClientRegisterContext) {
+                return getVoteForRolesMatched(((ClientUpdateContext)context).getToken());
+            } else {
+                throw new ClientPolicyException(OAuthErrorException.SERVER_ERROR, "unexpected context type.");
+            }
+        case UPDATE:
+            if (context instanceof AdminClientUpdateContext) {
+                return getVoteForRolesMatched(((ClientUpdateContext)context).getAuthenticatedUser());
+            } else if (context instanceof DynamicClientUpdateContext) {
+                return getVoteForRolesMatched(((ClientUpdateContext)context).getToken());
+            } else {
+                throw new ClientPolicyException(OAuthErrorException.SERVER_ERROR, "unexpected context type.");
+            }
+        default:
+            return ClientPolicyVote.ABSTAIN;
+        }
+    }
+
+    private ClientPolicyVote getVoteForRolesMatched(UserModel user) {
+        if (isRolesMatched(user)) return ClientPolicyVote.YES;
+        return ClientPolicyVote.NO;
+    }
+
+    private ClientPolicyVote getVoteForRolesMatched(JsonWebToken token) {
+        if (token == null) return ClientPolicyVote.NO;
+        if(isRoleMatched(token.getSubject())) return ClientPolicyVote.YES;
+        return ClientPolicyVote.NO;
+    }
+
+    private boolean isRoleMatched(String subjectId) {
+        if (subjectId == null) return false;
+        return isRolesMatched(session.users().getUserById(subjectId, session.getContext().getRealm()));
+    }
+
+    private boolean isRolesMatched(UserModel user) {
+        if (user == null) return false;
+
+        Set<String> expectedRoles = instantiateRolesForMatching();
+        if (expectedRoles == null) return false;
+
+        // user.getRoleMappingsStream() never returns null according to {@link UserModel.getRoleMappingsStream}
+        Set<String> roles = user.getRoleMappingsStream().map(RoleModel::getName).collect(Collectors.toSet());
```

Comment:
It seems that this won't take composite roles into account? Perhaps just "user.hasRole" is better as it takes composite roles and also groups into account. For example case when user is member of group "g" and this group has role "foo" assigned, then the user is considered to be in role "foo" due the indirection through the group.

Reply:
Yes. I understood the point not only treating the ordinal role but also the composit role. I'll use `user.hasRole` instaed of `user.getRoleMappingsStream`.